gpMgmt/bin/gppylib/db/dbconn.py

```diff
@@ -212,30 +203,40 @@ def connect(dburl, utility=False, verbose=False,
                 continue
             raise
 
-    if cnx is None:
+    if conn is None:
         raise ConnectionError('Failed to connect to %s' % dbbase)
 
     # NOTE: the code to set ALWAYS_SECURE_SEARCH_PATH_SQL below assumes it is not part of an existing transaction
-    conn = pgdb.pgdbCnx(cnx)
-
     #by default, libpq will print WARNINGS to stdout
     if not verbose:
-        cursor=conn.cursor()
-        cursor.execute("SET CLIENT_MIN_MESSAGES='ERROR'")
-        conn.commit()
-        cursor.close()
+        try:
+            cursor=conn.cursor()
+            cursor.execute("SET CLIENT_MIN_MESSAGES='ERROR'")
+            conn.commit()
+            cursor.close()
+        except Exception as e:
+            conn.close()
+            raise e
 
     # set client encoding if needed
     if encoding:
-        cursor=conn.cursor()
-        cursor.execute("SET CLIENT_ENCODING='%s'" % encoding)
-        conn.commit()
-        cursor.close()
+        try:
+            cursor=conn.cursor()
+            cursor.execute("SET CLIENT_ENCODING='%s'" % encoding)
+            conn.commit()
+            cursor.close()
+        except Exception as e:
+            conn.close()
+            raise e
```

Comment:
Why not add a wrapper that will always do this automatically, instead of doing it manually in many places?